sycl/doc/SYCLPluginInterface.md

```diff
@@ -41,12 +40,45 @@ once before any actual offload is attempted.
 
 ### Plugin discovery
 
-Plugins are physically dynamic libraries stored somewhere in the system where
-the SYCL runtime runs. TBD - design and describe the process in details.
+Plugins are physically dynamic libraries or shared objects.
+The process to discover plugins will follow the following guidelines.
+
+The SYCL Runtime will search for plugins at env LD_LIBRARY_PATH location on
+Linux and env LIB on Windows, with names in the format of libpiXXX.so and
+query them. An extension to the search mechanism is to use a configuration file
+which lists all the available plugins and their locations. 
+This file can be stored along with SYCL Runtime.
+Plugins should expose the information of supported PI API version as a constant
+string, so it can be read without loading the library.
+The user can select/disable a specifc plugin with an environment variable.
+The Plugin Interface queries the plugins on the supported PI version and check
+for compatibility.
+The Plugin Interface then queries each plugin for all Interface Functions
+corresponding function pointers and populate a list of these pointers for
+each plugin. As an extension, the plugin can provide a populated function
+pointer table/struct for each device attached to the plugin.
```

Comment:
In my view, organizing PI APIs into a table/structure is not natural for this project. This adds some additional coding complexities and constraints (like bothering about backward compatibility) but is not giving much in return except for multiple (<100) dlsym() calls needed, which I don't believe are going to be a bottleneck.

Note, there is already a loop in the pi.cpp over all the names in pi.def to do the bindings of all the PI interfaces.

Reply:
IIRC, there was a proposal from @keryell to use C++ feature (e.g. abstract class) to define PI interface.

Reply:
> This adds some additional coding complexities and constraints (like bothering about backward compatibility

I'd say on the contrary, interface table in single pointer simplifies backward compatibility, provided that all new APIs are added at the end. E.g. runtime can easily cast an older interface table to newer after checking the supported version in few lines of code.

> but is not giving much in return except for multiple (<100) dlsym() calls needed, which I don't believe are going to be a bottleneck.

saving dlsym is just marginal benefit. The main is better robustness, as this is an example of the encapsulation principle. A plugin offers the API as a package, not sycl.so picking what it wants. Also sycl.dll will manipulate entire PI implementations rather than individual functions/

Reply:
The current SYCL RT already implements the binding. Take a look here: https://github.com/intel/llvm/blob/f4d182ff8db947206af290ab6837981ff71a4bb5/sycl/source/detail/pi.cpp#L64-L67

I can agree to a different representation but there should be a good reason.
Knowing the layout of the provided "package" requires more hand-shaking for sure, and some maintenance cost too. What would be the benefits in return?

Reply:
> Knowing the layout of the provided "package" requires more hand-shaking for sure

no handshaking is needed. pi.h will define a
```
struct PI_Interface {
  int (*getPIVersion)();
  pi_result (*piPlatformsGet)(
    pi_uint32      num_entries,
    pi_platform *  platforms,
    pi_uint32 *    num_platforms);
    ...
};
```
and all plugins will #include and implement it. There are few tweaks needed to implement backward compatibility in a convenient way, as C does not allow inheritance, but that is minor issue.

> I can agree to a different representation but there should be a good reason.

I mentioned a few above. There should also be a good reason not to do so :)

Reply:
> IIRC, there was a proposal from @keryell to use C++ feature (e.g. abstract class) to define PI interface.

This would break the requirement to allow C plugins. Otherwise it would be very good.
Actually what I suggest above with PI_Interface kind of emulates this proposal in C (to some extent).

Reply:
Are you guys expecting to have C plugins? If the only use case is plugins for different SYCL backends, you may as well write their interface in C++ and avoid the function-pointer and dispatch tables which can sometimes be difficult to debug or understand.

Reply:
+1. OpenCL has C interfaces, but I don't see why this should affect SYCL runtime plug-in interface.

Reply:
The problem I see in a C-based plug-in approach is that SYCL is written in modern C++. And this is good for humankind. :-)
So any dynamic dispatch is already provided natively in C++ with dynamic polymorphism.
So, normally, it should be easy to plug in another C++ run-time in the SYCL run-time by writing a light C++ shim.
If you want plug-ins written in some other languages (Fortran, Cobol, C...) you write some other C++ shims that will do some façade adaptation, for example calling some C functions as it is mainly done in the current PI implementation.

By providing only a C-plugin interface:
- if you target a C++ runtime, you need to wrap it up into a C low-level interface targeting the C façade of the plugin, instead of a direct C++ light-weight shim :-( ;
- if you have a C runtime, you need anyway to implement another C shim to adapt it to the plug-in interface because the probability is 0 that your C runtime implement exactly what you want in the PI interface, so it is not simpler than to write directly a C++ adapter to this new C runtime...

Reply:
If nobody sees any value in supporting C plugins, I'm OK with C++ interface. We should probably have some restrictions for future extensions, like not using string or STL containers in the signatures to provide some compatibility between C++ runtime flavors.

Reply:
The major problem I see with C++ interfaces is that it's ABI is not as mush standard as C one. There are multiple compatibility issues with C++ ABI, which are also different for different OS. And there is way more robust integration possible for C plugin with a not C/C++ layer. We've had this discussion before and after that we already implemented the C interfaces. Let's stick to C for across the binaries interfaces.

The real issue being raised here, I believe, is to whether (1) export the entire collection of interfaces as a single object, or (2) export them individually without bundling into a super-structure. OpenCL is doing #1 (and @kbobrovs is voting for that too), my voice is for #2 for the reasons explained above. Any other preferences on this small aspect?

Reply:
Probably we could follow the same approach OpenCL ICD uses: having super-structure for all required APIs and exporting one by one for extensions.

Reply:
There are some interesting examples in the Boost.DLL tutorial http://apolukhin.github.io/Boost.DLL/
A simple C++ design could be just 2 symbols exported by a DLL, 1 plugin factory function + 1 destructor function. Or the last one could be hidden in the unload hook.
The C++ instance returned by the factory provides everything, the member functions providing each PI operations but also some member functions to know for each PI operation if the member function does something (i.e. the operation is implemented by the plugin) or not. Instead of 1 querying member function per operation member function, it could be a big `constexpr` Boolean table instead, storing which operation does actually something.

Reply:
With the necessity to de-multiplex between multiple PI plugins (explained here https://github.com/intel/llvm/pull/680#discussion_r332278399), I no longer think we should be doing #2 above. Switching between the tables is so much easier and more natural, and we shall do that.

Reply:
We shall be adding new PI API interface(s) that each PI plugin should implement and that would return the PI plugin descriptor structure containing [at least] the following: version information, supported architectures, and a dispatch table with all the supported PI API implementations.

Reply:
We agree with the decision to stick with C for the API. It lets us map very closely to the underlying API that is being abstracted while exposing simple opaque types and functions to interface with the underlying API. OpenCL can currently be directly passed through. Additionally, we agree with @smaslov-intel  that C is much better for ABI stability than C++. This is why C is used so often for plugin style libraries. Additionally, if the PI API is ever going to be used in other languages or projects, the C API will make this much much easier to create language bindings.

Reply:
In embedded systems, such as in ADAS, the environment is completely defined and even sometimes certified. So, this plug-in story is not fundamental for us, it is unlikely we just discover at the last second what will happen (when you brake in your car... :-) ). Thus I do not have a strong opinion abut this full story. A more typical use-case I can see would be to modify directly in C++ the SYCL run-time and the compiler flow for a given target. And then we do not really care about dynamic library loading at run-time and low-level C pointer atrocities... :-)

Reply:
More thoughts why I prefer a C -based interface:
* If the PI API stays compatible with OpenCL, then functions pointers can directly be wired up (like the current code does) - regardless of if a lookup table is used or not.
* My understanding of a C++ approach is as follows:
  * A virtual function call is required which then either
    * needs to call a stored OpenCL API function pointer, or
    * calls an OpenCL API through the ICD mechanism (another function pointer lookup indirection).
  * Due to the double indirection this sounds more expensive than the C approach to me.
* I have no experience with ABI stability and versioning based on abstract C++ base classes.